src/modules/roc_sdp/session_description.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2019 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include "roc_sdp/session_description.h"
+
+namespace roc {
+namespace sdp {
+
+SessionDescription::SessionDescription(core::IAllocator& allocator)
+    : guid_(allocator)
+    , media_descriptions_()
+    , allocator_(allocator) {
+    clear();
+}
+
+void SessionDescription::clear() {
+    guid_.clear();
+    origin_unicast_address_.clear();
+    session_connection_data_.clear();
+
+    while (media_descriptions_.size() > 0) {
+        core::SharedPtr<MediaDescription> m = media_descriptions_.back();
+        media_descriptions_.remove(*m);
+    }
+}
+
+const char* SessionDescription::guid() const {
+    if (guid_.is_empty()) {
+        return NULL;
+    }
+    return guid_.c_str();
+}
+
+bool SessionDescription::set_guid(const char* start_p_origin_username,
+                                  const char* end_p_origin_sess_id,
+                                  const char* start_p_origin_nettype,
+                                  const char* end_p_origin_addr) {
+    core::StringBuilder b(guid_.raw_buf());
+
+    if (!b.append_str_range(start_p_origin_username, end_p_origin_sess_id)) {
+        return false;
+    }
+
+    b.append_char(' ');
```

Comment:
Yes we can! What would you use ?
In the RFC they don't really talk about the guid (or I missed it ?), except in the quote from the previous comment.

And for the parameters of set_guid I would have an start_p and end_p for each variable if we do a non-space delimiter.

Reply:
To produce GUIDs unambiguously, we should use a delimiter that is prohibited from use in GUID fields. So we can use any character that can't be used in `o=` sub-fields (username, sess-id, etc). Space is good, but it would be more convenient if GUIDs didn't contain spaces. Are there any other characters prohibited by RFC in `o=` sub-fields? Preferably, some printable ASCII character.

Reply:
In the origin field, a username is a non-ws-string, which is defined by:  1*(VCHAR/%x80-FF) = string of visible characters.

The [RFC 4566](https://tools.ietf.org/html/rfc4566#section-9) comments on the username field: "pretty wide definition, but doesn't include space".

In [RFC 5234](https://tools.ietf.org/html/rfc5234), they define VCHAR = %x21-7E.

So except a space, I don't think there is any ASCII character that is available :/

Reply:
Thanks, I see. Then let's keep using space as delimiter.